src/Physics/Hadronization/LeptoHadPythia6.cxx

```diff
@@ -0,0 +1,508 @@
+      double pT  = fRemnantPT * TMath::Sqrt( -1*TMath::Log( rnd->RndHadro().Rndm() ) );
+      double pT2 = TMath::Power(pT,2);
+      double pr  = TMath::Power(m_hadron,2)+pT2;
+      //to generate the longitudinal scaling variable z in jet fragmentation using PYTHIA function
+      // Split energy-momentum of remnant using PYTHIA function
+      // z=E-pz fraction for rema forming jet-system with frag_q
+      // 1-z=E-pz fraction for hadron
+      double z;
+      int kfl1 = 1;
+      int kfl3 = 0;
+      pyzdis_(&kfl1,&kfl3,&pr,&z);
+
+      // Energy of trasnfered to the hadron
+      double tm_hadron = pr / z / W;
+      double E_hadron   = 0.5 * ( z*W + tm_hadron );  //E_hadron - pz = zW
+      double E_pz       = W - tm_hadron;
+      double WT         = (1-z) * W * E_pz - pT2;
+
+      // Check if energy in jet system is enough for fragmentation.
+      if ( WT > TMath::Power(m_frag+m_rema+fMinESinglet,2) ) {
+
+        // Energy of transfered to the fragmented quark and rema system
+        // Applying energy conservation
+        WT = TMath::Sqrt( WT + pT2 );
+        double tm_rema   = TMath::Power(m_rema,2) + pT2;
+        double E_frag    = 0.5 * ( WT + ( TMath::Power(m_frag,2) - tm_rema)/WT ); //E_frag + E_rema = WT
+        double E_rema    = 0.5 * ( WT + (-TMath::Power(m_frag,2) + tm_rema)/WT );
+        double x_rema    = -1 * TMath::Sqrt( TMath::Power(E_rema,2) - tm_rema );
+        double theta_rema;
+        theta_rema = pyangl_(&x_rema,&pT);
+
+        // Select a phi angle between between particles randomly
+        double phi = 2*kPi*rnd->RndHadro().Rndm();
+
+        double dbez = (E_pz-(1-z)*W)/(E_pz+(1-z)*W);
+        double pz_hadron  = -0.5 * ( z*W - tm_hadron );
+
+        // Input the three particles to PYTHIA in the CM frame
+        // If a top quark is produced we decay it because it does not hadronize
+
+        fPythia->Py1ent( -1, frag_quark, E_frag, 0.,         0. );           //k(1,2) = 2
+        if (TMath::Abs(frag_quark) > 5 ) {
+          int ip = 1;
+          pydecy_(&ip);
+        }
+        fPythia->Py1ent( fPythia->GetN()+1, rema, E_rema, theta_rema, phi ); //k(2,2) = 1
+
+        int imin     = 0;
+        int imax     = 0;
+        double the  = 0.; double ph   = 0.;
+        double dbex = 0.; double dbey = 0.; 
+        pyrobo_( &imin , &imax, &the, &ph, &dbex, &dbey , &dbez );
+        double theta_hadron = pyangl_(&pz_hadron,&pT);
+
+        fPythia->SetMSTU( 10, 1 ); //keep the mass value stored in P(I,5), whatever it is.
+        fPythia->SetP( fPythia->GetN()+1, 5, m_hadron );
+        fPythia->Py1ent( fPythia->GetN()+1, hadron, E_hadron, theta_hadron, phi + kPi );
+        fPythia->SetMSTU( 10, 2 ); //find masses according to mass tables as usual.
+
+        // Target remnants required to go backwards in hadronic cms
+        if ( fPythia->GetP(fPythia->GetN()-1,3)<0 && fPythia->GetP(fPythia->GetN(),3)<0 ) break; //quit the while from line 368
```

Comment:
This combination of while loops and if's is a bit hard to follow, I'll assume this is right.   But I'm guessing that the comment after the `break` isn't right about line number any more.  I'd suggest that rather than using the line number it give the actual while conditional.